src/app/AttributePathExpandIterator-DataModel.cpp

```diff
@@ -0,0 +1,243 @@
+    case Clusters::Globals::Attributes::EventList::Id:
+#endif // CHIP_CONFIG_ENABLE_EVENTLIST_ATTRIBUTE
+        return Clusters::Globals::Attributes::AttributeList::Id;
+    case Clusters::Globals::Attributes::AttributeList::Id:
+        return std::nullopt;
+    }
+
+    AttributeEntry entry = mDataModel->NextAttribute(mOutputPath);
+    if (entry.IsValid())
+    {
+        return entry.path.mAttributeId;
+    }
+    return Clusters::Globals::Attributes::GeneratedCommandList::Id;
+}
+
+std::optional<ClusterId> AttributePathExpandIteratorDataModel::NextClusterId()
+{
+
+    if (mOutputPath.mClusterId == kInvalidClusterId)
+    {
+        if (mpAttributePath->mValue.HasWildcardClusterId())
+        {
+            ClusterEntry entry = mDataModel->FirstCluster(mOutputPath.mEndpointId);
+            return entry.IsValid() ? std::make_optional(entry.path.mClusterId) : std::nullopt;
+        }
+
+        // only return a cluster if it is valid
+        const ConcreteClusterPath clusterPath(mOutputPath.mEndpointId, mpAttributePath->mValue.mClusterId);
+        if (!mDataModel->GetClusterInfo(clusterPath).has_value())
+        {
+            return std::nullopt;
+        }
+
+        return mpAttributePath->mValue.mClusterId;
+    }
+
+    VerifyOrReturnValue(mpAttributePath->mValue.HasWildcardClusterId(), std::nullopt);
+
+    ClusterEntry entry = mDataModel->NextCluster(mOutputPath);
+    return entry.IsValid() ? std::make_optional(entry.path.mClusterId) : std::nullopt;
+}
+
+std::optional<ClusterId> AttributePathExpandIteratorDataModel::NextEndpointId()
+{
+    if (mOutputPath.mEndpointId == kInvalidEndpointId)
+    {
+        if (mpAttributePath->mValue.HasWildcardEndpointId())
+        {
+            EndpointId id = mDataModel->FirstEndpoint();
+            return (id != kInvalidEndpointId) ? std::make_optional(id) : std::nullopt;
+        }
+
+        return mpAttributePath->mValue.mEndpointId;
+    }
+
+    VerifyOrReturnValue(mpAttributePath->mValue.HasWildcardEndpointId(), std::nullopt);
+
+    EndpointId id = mDataModel->NextEndpoint(mOutputPath.mEndpointId);
+    return (id != kInvalidEndpointId) ? std::make_optional(id) : std::nullopt;
+}
+
+void AttributePathExpandIteratorDataModel::ResetCurrentCluster()
+{
+    ChipLogError(Test, "Reset current cluster data model");
+
+    // If this is a null iterator, or the attribute id of current cluster info is not a wildcard attribute id, then this function
+    // will do nothing, since we won't be expanding the wildcard attribute ids under a cluster.
+    VerifyOrReturn(mpAttributePath != nullptr && mpAttributePath->mValue.HasWildcardAttributeId());
+
+    // Reset to the first attribute of the current cluster.
+    // It is expected that the output path is already positioned on a proper cluster
+    //
+    AttributeEntry entry = mDataModel->FirstAttribute(mOutputPath);
+    VerifyOrDie(entry.IsValid()); // TODO: is this sane? do we require validity here?
```

Comment:
I don't understand this block of code.  Is this just trying to reset the data model's optimization cache bits?  Why?

Reply:
Removed this double-check as our "next" implementation has all the logic existing regarding selecting the first valid cluster or making `Get` return false.